traffic_ops/traffic_ops_golang/server/servers.go

```diff
@@ -322,6 +325,27 @@ func (sv *TOServer) Update() (error, error, int) {
 	if sv.IP6Address != nil && len(strings.TrimSpace(*sv.IP6Address)) == 0 {
 		sv.IP6Address = nil
 	}
+
+	// see if type changed
+	typeID := -1
+	if err := sv.APIInfo().Tx.QueryRow("SELECT type FROM server WHERE id = $1", sv.ID).Scan(&typeID); err != nil {
+		if err == sql.ErrNoRows {
+			return errors.New("no server found with this id"), nil, http.StatusNotFound
+		}
+		return nil, fmt.Errorf("getting current server type: %v", err), http.StatusInternalServerError
+	}
+
+	// If type is changing ensure it isnt assigned to any DSes
+	if typeID != *sv.TypeID {
+		dsIDs := []int64{}
+		if err := sv.APIInfo().Tx.QueryRowx("SELECT ARRAY(SELECT deliveryservice FROM deliveryservice_server WHERE server = $1)", sv.ID).Scan(pq.Array(&dsIDs)); err != nil && err != sql.ErrNoRows {
+			return nil, fmt.Errorf("getting server assigned delivery services: %v", err), http.StatusInternalServerError
+		}
+		if len(dsIDs) != 0 {
+			return errors.New("server type can not be updated when it is currently assigned to delivery services"), nil, http.StatusBadRequest
```

Comment:
Personally I like [`409 Conflict`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) for things like this. Do with that what you will.